src/Disks/IDisk.cpp

```diff
@@ -27,24 +29,75 @@ void copyFile(IDisk & from_disk, const String & from_path, IDisk & to_disk, cons
     copyData(*in, *out);
 }
 
-void IDisk::copy(const String & from_path, const std::shared_ptr<IDisk> & to_disk, const String & to_path)
+
+using ResultsCollector = std::vector<std::future<void>>;
+
+void asyncCopy(IDisk & from_disk, String from_path, IDisk & to_disk, String to_path, Executor & exec, ResultsCollector & results)
 {
-    if (isFile(from_path))
+    if (from_disk.isFile(from_path))
     {
-        DB::copyFile(*this, from_path, *to_disk, to_path + fileName(from_path));
+        auto result = exec.execute(
+            [&from_disk, from_path, &to_disk, to_path]()
+            {
+                setThreadName("DiskCopier");
+                DB::copyFile(from_disk, from_path, to_disk, to_path + fileName(from_path));
+            });
+
+        results.push_back(std::move(result));
     }
     else
     {
         Poco::Path path(from_path);
         const String & dir_name = path.directory(path.depth() - 1);
         const String dest = to_path + dir_name + "/";
-        to_disk->createDirectories(dest);
+        to_disk.createDirectories(dest);
 
-        for (auto it = iterateDirectory(from_path); it->isValid(); it->next())
+        for (auto it = from_disk.iterateDirectory(from_path); it->isValid(); it->next())
+            asyncCopy(from_disk, it->path(), to_disk, dest, exec, results);
+    }
+}
+
+void IDisk::copy(const String & from_path, const std::shared_ptr<IDisk> & to_disk, const String & to_path)
+{
+    auto exec = to_disk->getExecutor();
+    ResultsCollector results;
+
+    asyncCopy(*this, from_path, *to_disk, to_path, *exec, results);
+
+    for (auto & result : results)
+        result.wait();
+    for (auto & result : results)
+        result.get();
+}
+
+/// Executes task synchronously in case when disk doesn't support async operations.
+class SyncExecutor : public Executor
+{
+public:
+    SyncExecutor() = default;
+    std::future<void> execute(std::function<void()> task) override
+    {
+        auto promise = std::make_shared<std::promise<void>>();
+        try
+        {
+            task();
+            promise->set_value();
+        }
+        catch (...)
         {
-            copy(it->path(), to_disk, dest);
+            try
+            {
+                promise->set_exception(std::current_exception());
+            }
+            catch (...) { }
         }
+        return promise->get_future();
     }
+};
+
+std::unique_ptr<Executor> IDisk::getExecutor()
+{
+    return std::make_unique<SyncExecutor>();
```

Comment:
So this method passes ownership to the caller? But why, isn't `Executor` supposed to be some mechanism of `Disk` that has the same lifetime? Why do we have to create a new object on each call?

Also, both of your implementations of executor are stateless, and chosen statically, not at runtime. The default implementation is synchronous, and S3 always uses asynchronous. Why do you even need an interface for this, maybe it can just be a virtual function?

Reply:
Executor interface is re-usable while virtual function is not. E.g. AsyncExecutor can be used for some other disks like `HDFS`. That was my main concern to make a separate interface to execute tasks.

Reply:
OK, let's remove the allocation then, other than that everything looks good.

Reply:
Done